### PR TITLE
Main

### PR DESCRIPTION
**Pull request by onedebos**

### CODE REVIEW

**Comment by onedebos** [comment]
Ruby Tic-tac-toe project.completed.

File: lib/game.rb [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  def initialize
+    @board = [' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ']
+    @player1 = Player.new('Player 1', 'X')
+    @player2 = Player.new('Player 2', 'O')
+    @current_player = @player1
+    @end_game = false
+  end
+
+  def draw_board
+    puts "\n"
```
Comment (bellom) [inline lib/game.rb]: `main.rb` file should be responsible for user interaction. This is the only one place where puts and get.chomp can be used.

File: lib/game.rb [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    puts " #{@board[0]} | #{@board[1]} | #{@board[2]}"
+    puts '-----------'
+    puts " #{@board[3]} | #{@board[4]} | #{@board[5]}"
+    puts '-----------'
+    puts " #{@board[6]} | #{@board[7]} | #{@board[8]}"
+  end
+
+  def draw_legend
+    puts 'Legend'
```
Comment (bellom) [inline lib/game.rb]: `main.rb` file should be responsible for user interaction. This is the only one place where puts and get.chomp can be used._

File: lib/game.rb [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  end
+
+  def which_player?
+    @current_player = @current_player == @player1 ? @player2 : @player1
+  end
+
+  def player_input
+    puts "Your turn, #{@current_player.name}. Pick a cell by
+    choosing a number between 1-9."
+    input = gets.chomp.to_i
```
Comment (bellom) [inline lib/game.rb]: `main.rb` file should be responsible for user interaction. This is the only one place where puts and get.chomp can be used._

File: lib/game.rb [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    if input.between?(1, 9)
+      input - 1
+    else
+      puts ' You need to enter a number between 1 and 9.'
```
Comment (bellom) [inline lib/game.rb]: `main.rb` file should be responsible for user interaction. This is the only one place where puts and get.chomp can be used._

File: lib/game.rb [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    end
+  end
+
+  def player_move
+    move = player_input
+    until @board[move] == ' '
+      puts 'This cell is not available.'
```
Comment (bellom) [inline lib/game.rb]: `main.rb` file should be responsible for user interaction. This is the only one place where puts and get.chomp can be used._

File: lib/game.rb [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    end
+    @board[move] = @current_player.symbol
+    @current_player.fill_cells << move
+  end
+
+  def player_wins?
+    conditions = [[0, 1, 2], [3, 4, 5], [6, 7, 8], [0, 3, 6], [1, 4, 7], [2, 5, 8], [0, 4, 8], [2, 4, 6]]
+    conditions.each do |condition|
+      if (condition - @current_player.fill_cells).empty?
+        puts "#{@current_player.name} wins!"
```
Comment (bellom) [inline lib/game.rb]: `main.rb` file should be responsible for user interaction. This is the only one place where puts and get.chomp can be used._

File: lib/game.rb [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        @end_game = true
+      end
+    end
+    if @player1.fill_cells.size + @player2.fill_cells.size >= 9
+      puts "It's a draw."
```
Comment (bellom) [inline lib/game.rb]: `main.rb` file should be responsible for user interaction. This is the only one place where puts and get.chomp can be used._

**Review by bellom — CHANGES_REQUESTED** [review]
Hi Team,
Great work guys, game well played and nice job using `Git Flow` but you didn't go through all the instructions from pathwright.
_Always read the pathwright instructions before starting a project to avoid multiple code review request, according to the pathwright, `main.rb` file should be responsible for user interaction. This is the only one place where puts and get.chomp can be used._

_Kindly re-submit for review after making all those changes slightly_smiling_face Thumps -up await you._

**Review by AdnanAfsari — CHANGES_REQUESTED** [review]
Hello @onedebos @vieiramanda11 ✋ 

Nice work guys but i think you are yet to make some changes the previous reviewer requested to change, those changes are not big i guess.
Please go ahead and make current those changes and submit for anther review 😄 

Happy Coding

File: lib/game.rb [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    end
+    @board[move] = @current_player.symbol
+    @current_player.fill_cells << move
+  end
+
+  def player_wins?
+    conditions = [[0, 1, 2], [3, 4, 5], [6, 7, 8], [0, 3, 6], [1, 4, 7], [2, 5, 8], [0, 4, 8], [2, 4, 6]]
+    conditions.each do |condition|
+      if (condition - @current_player.fill_cells).empty?
+        did_win
```
Comment (mosaaleb) [inline lib/game.rb]: Applying SRP here we can refactor this method into smaller method each does only one thing. As you can see this method:
1. checks for winning.
2. checks for draw.
3. print a message.

We can refactor it like this:
```ruby
def end_game?
    player_wins? || game_draw?
end
```
then we need to implement these methods accordingly.

Comment (mosaaleb) [inline lib/game.rb]: Applying SRP here we can refactor this method into smaller method each does only one thing. As you can see this method:
1. checks for winning.
2. checks for draw.
3. print a message.

We can refactor it like this:
```ruby
def end_game?
    player_wins? || game_draw?
end
```
then we need to implement these methods accordingly.

**Review by mosaaleb — CHANGES_REQUESTED** [review]
### Well done team :clap:

Yet there are some work we need to do before moving on:

- [ ] When a player enters a number not in (1..9) the application throws an error and exit execution, it need to repeat the process without exiting.

- [ ] The project should have descriptive readme file.

- [ ] (`main.rb` file should be responsible for user interaction) doesn't mean that you wrap the puts and gets inside method and still use them in other classes. `Game` class should not do any kind of interactions even if it is wrapped inside another methods. 
The idea behind this comes from single responsibility principle whereas each class and each method should only do one thing. For example `Game` class should decide if we have reached a winning position or not, then the `main.rb` can use this functionality to `put` a relevant message. Now we have separated the functionality of each class. 

What you have done from the previous review is just moving the interaction messages from one class to a module.

A good indication of applying a single responsibility principle (SRP) is to ask yourself what this `Game` class do, if you said the word (and) then it does too much. In our case `Game` class should only decide the workflow of the game and this can be represnted in multiple methods: `winning?` `draw?` `change_turns` and so on. Then for our class in `main.rb` we need to instantiate a `Game` object and always checking for these methods and hence decide what message we should print to the user.

Please handle these requirements and submit a new code review.

#### Happy Coding :computer:

**Review by maelfosso — APPROVED** [review]
# Good job :+1: 

You can merge with the `development` branch. 
Good luck with the next project.

`*OPTIONAL*` : Please remove the `Game` class from `main.rb` and create a simple class for it such that it doesn't have user interactions which must only be into `main.rb`. Otherwise, you will have some problem when writing test for that class in the next project.

**Pull request merged**